src/core/models.py

```diff
@@ -20,34 +22,64 @@ class StateBase:
     """Base state object."""
 
     def __init__(
-        self, relation: Relation | None, component: Unit | Application, substrate: SUBSTRATES
+        self,
+        relation: Relation | None,
+        data_interface: Data,
+        component: Unit | Application,
+        substrate: SUBSTRATES,
     ):
         self.relation = relation
+        self.data_interface = data_interface
         self.component = component
         self.substrate = substrate
 
-    @property
-    def relation_data(self) -> MutableMapping[str, str]:
-        """The raw relation data."""
-        if not self.relation:
-            return {}
+    def update(self, items: dict[str, str]) -> None:
+        """Changes the state."""
+        raise NotImplementedError
+
+    def data(self) -> MutableMapping:
+        """Data representing the state."""
+        raise NotImplementedError
+
+
+class RelationState(StateBase):
+    """Base state object."""
 
-        return self.relation.data[self.component]
+    def __init__(
+        self,
+        relation: Relation,
+        data_interface: Data,
+        component: Unit | Application,
+        substrate: SUBSTRATES,
+    ):
+        super().__init__(relation, data_interface, component, substrate)
+        # Redundant definition as lint can't resolve that super's relation may be None
+        self.relation = relation
+        self.relation_data = self.data_interface.as_dict(self.relation.id)
+
+    @property
+    def data(self) -> MutableMapping:
+        """Data representing the state."""
+        return self.relation_data
 
     def update(self, items: dict[str, str]) -> None:
         """Writes to relation_data."""
-        if not self.relation:
-            return
+        delete_fields = [key for key in items if not items[key]]
+        update_content = {k: items[k] for k in items if k not in delete_fields}
 
-        self.relation_data.update(items)
+        self.relation_data.update(update_content)
 
+        for field in delete_fields:
+            del self.relation_data[field]
 
-class ZKClient(StateBase):
+
+class ZKClient(RelationState):
     """State collection metadata for a single related client application."""
 
     def __init__(
         self,
-        relation: Relation | None,
+        relation: Relation,
```

Comment:
**todo:** Based on other comments, this will be type `Relation | None`

Reply:
Same as https://github.com/canonical/zookeeper-operator/pull/129#discussion_r1568894833

Reply:
Duplicate, done.